src/IO/ReadWriteBufferFromHTTP.h

```diff
@@ -214,20 +216,30 @@ namespace detail
                 }
                 catch (const Poco::Exception & e)
                 {
+                    if (i == settings.http_max_tries - 1)
+                        throw;
+
                     LOG_ERROR(log, "Failed to make HTTP_HEAD request to {}. Error: {}", uri.toString(), e.displayText());
                 }
             }
 
             if (response.hasContentLength())
-                read_range.end = getRangeBegin() + response.getContentLength();
+            {
+                if (!read_range.end)
+                    read_range.end = getRangeBegin() + response.getContentLength();
```

Comment:
Does this fix invalid range for parquet?
I'm not sure I follow what happened here.

Reply:
No. The previous code just does not seem correct to me, calling `getFileSize()` at random time could result in incorrect `read_range.begin` value (though in fact it is not called at random time, usually only once - before starting reading, but anyway not good).

Invalid range ([from this slack discussion](https://clickhouse-inc.slack.com/archives/C02F2LML5UG/p1657813291532319?thread_ts=1657812771.934939&cid=C02F2LML5UG)) was a result of incorrect logging -- stupid bug (there was `read_range.end ? *read_range.end : '-'));` but needed `read_range.end ? toString(*read_range.end) : "-",`). The actual range request was correct, I also checked that it was within the file size, e.g. no out of bound range. The point is -- even retries did not help with getting `http range not satisfiable` for me, I checked with 10 retries. But as stated in https://github.com/ClickHouse/product/issues/176 -- retries do help with such kind of error eventually, so that is why I made it possible to retry.

Reply:
ah nice, implicit casts are the best.
thanks for explaining